GitHub gist

dotps1 opened the pull request:
# Improvements / Enhancements

Added initial classes for GitHubGists which also lead to other classes that are partially embedded.
- GitHubGist
- GitHubGistFile
- GitHubGistOwner
- GitHubOwner (Derived from GitHubGistOwner)
- GitHubPlan
  From what I am seeing, we cannot have classes existing in more then one file.  (I tried to make a new file called 'GitHubOwners', and when I went to use the GitHubGistOwner class in the GitHubGist class, it errors saying that it could not find the class).

added a method to the GitHubGistFile method called GetFileContent allowing for the content of a file to be fetched if it is not present.  which seems to be any time a Gist is retrieved without using the Id.  This allows for the following syntax to work properly:

``` PowerShell
Get-GitHubGist | Select -First 1 | Save-GitHubGist
```

Updated Save-GitHubGist to accept object of type GitHubGist rather then System.Object.

I think that's it.  I didn't move any of the classes you have defined already, but that will need to be done, unless i'm missing something.

_this is my first attempt at classes, so feel free to be brutal with comments and suggestions lol_

PlagueHO commented on Classes/GitHubGist.ps1:
```diff
@@ -1,18 +0,0 @@
-        $this.Owner = $Owner;
-        $this.Description = $Description;
-    }
-}
```
Needs a newline char here.

PlagueHO commented on Classes/GitHubIssue.ps1:
```diff
@@ -15,7 +15,7 @@ class GitHubIssue {
+### In the GitHub object model, a Pull Request "is an" issue, but not vice versa
 class GitHubPullRequest : GitHubIssue {
     
 }
```
Needs a newline char here.

PlagueHO commented on Classes/PSGitHub.Classes.ps1:
```diff
@@ -0,0 +1,206 @@
+        $this.Files = $Files
+        $this.Public = $Public
+    }
+}
```
Needs a newline char here.

PlagueHO commented on Functions/Public/Get-GitHubGist.ps1:
```diff
@@ -81,9 +81,7 @@ function Get-GitHubGist {
+    foreach ($result in (Invoke-GitHubApi @apiCall)) {
+        [GitHubGist]::new($result)
     }
 }
```
Needs a newline char here.

dotps1 commented:
I'm not following that comment on any of those.

PlagueHO commented:
Hi there @dotps1,
All PS1 files should end with a new line character. GitHub will show you if you've missed any:
![2016-03-31_16-42-29](https://cloud.githubusercontent.com/assets/7589164/14164834/a37ac920-f75f-11e5-8287-ccf963ad260d.png)
See the little red symbol at the bottom. I'm always doing this myself TBH.
Thanks! :smile:

dotps1 commented:
Why do they need that?  I've never done that on any of my stuff.

Sent from my iPhone

> On Mar 30, 2016, at 11:44 PM, Daniel Scott-Raynsford notifications@github.com wrote:
> 
> Hi there @dotps1,
> All PS1 files should end with a new line character. GitHub will show you if you've missed any:
> 
> See the little red symbol at the bottom. I'm always doing this myself TBH.
> Thanks! 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub

PlagueHO commented:
See: http://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

PlagueHO commented:
All of the MSFT PowerShell community repos require this as well, so it's generally a good habit to get into (although as I said I'm always missing it myself).

pcgeek86 commented on Classes/PSGitHub.Classes.ps1:
```diff
@@ -0,0 +1,257 @@
+}
+
+Class GitHubUser {
+    [String]$AvatarUrl
```
Maybe we could use `System.Uri` instead of "string" for some of these URL-related properties.

dotps1 replied:
totally agree.  done.

pcgeek86 commented on Classes/PSGitHub.Classes.ps1:
```diff
@@ -0,0 +1,257 @@
+    GitHubGistForks() { }
+}
+
+# Object may not be only used for Gists.
```
Explain why??

dotps1 replied:
I'm not sure what you asking, but I assume its the comment.  Basically the History, Fork, and ChangeStatus I have a feeling will be used in repos as well as Gists, if that is the case, I think we should name them better.  I added some better documentation to express this.

dotps1 commented:
I noticed my GitHubGistFork class was missing a few properties, I corrected that, also, GitHubGistHistory no longer derives that, no point.  I added a lot of documentation to the Classes.  I also added new lines at the end of all the files I have worked on.

dotps1 commented:
I added the Copy-GitHubGist function with an alias of Fork-GitHubGist.  I also noticed my constructor name was mistyped for GitHubGistFork, I corrected that, I also added a try catch to the private method Invoke-GitHubApi; if it fails it will throw and error containing the message property from the api response.  this looks much better then the json response.

pcgeek86 commented on PSGitHub.psm1:
```diff
@@ -27,5 +27,11 @@ foreach ($File in $FunctionList) {
 
+#region aliases
+
+New-Alias -Name Fork-GitHubGist -Value Copy-GitHubGist
```
Let's define aliases in a separate file, and load those. I'd suggest creating an `\Aliases` subdirectory.

felixfbecker commented:
Closing since we have these cmdlets in PSGitHub now.
